spherical_splines/SHTOOLS/man/man1/plmbar_d1.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PLMBAR_D1 1"
.TH PLMBAR_D1 1 "2009-08-18" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "PlmBar_d1"
.IX Header "PlmBar_d1"
.IP "PlmBar_d1 \-" 12
.IX Item "PlmBar_d1 -"
Compute all the geodesy-normalized associated Legendre functions and first derivatives.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 PlmBar_d1 (" 23
.IX Item "SUBROUTINE PlmBar_d1 ("
P, \s-1DP\s0, \s-1LMAX\s0, Z, \s-1CSPHASE\s0, \s-1CNORM\s0 )
.RS 4
.IP "REAL*8" 20
.IX Item "REAL*8"
P((\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2), \s-1DP\s0((\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2), Z
.IP "\s-1INTEGER\s0" 20
.IX Item "INTEGER"
\&\s-1LMAX\s0
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 20
.IX Item "INTEGER, OPTIONAL"
\&\s-1CSPHASE\s0, \s-1CNORM\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIPlmBar_d1\fR will calculate all of the geodesy-normalized associated Legendre functions and first derivatives up to degree \s-1LMAX\s0 for a given argument. These are calculated using a standard three-term recursion formula, and in order to prevent overflows, the scaling approach of Holmes and Featherstone (2002) is utilized. These functions are accurate to about degree 2800. The index of the array corresponding to a given degree L and angular order M can be computed by a call to \fIPlmIndex\fR.
.PP
The integral of the squared Legendre functions over the interval [\-1, 1] is 2 * (2 \- delta(0,m)), where delta is the Kronecker delta function. If the optional parameter \s-1CNORM\s0 is set equal to 1, the complex normalization will be used where the integral of the squared Legendre functions over the interval [\-1, 1] is 2. The default is to exclude the Condon-Shortley phase, but this can be modified by setting the optional argument \s-1CSPHASE\s0 to \-1. Note that the derivative of the Legendre functions is calculated with respect to its arguement Z, and not latitude or colatitude. If Z=cos(theta), where theta is the colatitude, then it is only necessary to multiply \s-1DP\s0 by \-sin(theta) to obtain the derivative with respect to theta.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "P" 10
.IX Item "P"
(output) REAL*8, \s-1DIMENSION\s0 ((\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2)
.Sp
An array of geodesy-normalized associated Legendre functions up to degree \s-1LMAX\s0. The index corresponds to L*(L+1)/2 + M + 1, which can be calculated by a call to \fIPlmIndex\fR.
.IP "\s-1DP\s0" 10
.IX Item "DP"
(output) REAL*8, \s-1DIMENSION\s0 ((\s-1LMAX+1\s0)*(\s-1LMAX+2\s0)/2)
.Sp
An array of the first derivatives of the geodesy-normalized associated Legendre functions up to degree \s-1LMAX\s0. The index corresponds to L*(L+1)/2 + M + 1, which can be calculated by a call to \fIPlmIndex\fR.
.IP "\s-1LMAX\s0" 10
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum degree of the associated Legendre functions to be computed. If \s-1LMAX\s0 is \-1, allocated memory will be deallocated (see notes below).
.IP "Z" 10
.IX Item "Z"
(input) REAL*8
.Sp
The argument of the associated Legendre functions.
.IP "\s-1CSPHASE\s0" 10
.IX Item "CSPHASE"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
If 1 (default), the Condon-Shortley phase will be excluded. If \-1, the Condon-Shortley phase of (\-1)^m will be appended to the associated Legendre functions.
.IP "\s-1CNORM\s0" 10
.IX Item "CNORM"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
If 1, the complex normalization of the associated Legendre functions will be used. The default is to use the real normalization.
.SH "NOTES"
.IX Header "NOTES"
This routine saves the three-term recursion factors and square roots of the integers the first time being called. If subsequent calls possess the same value of \s-1LMAX\s0, these will not be recomputed. If you wish to deallocate this memory, which is an array of length (\s-1LMAX+1\s0)*(\s-1LMAX+2\s0), recall this routine with LMAX=\-1.
.SH "REFERENCES"
.IX Header "REFERENCES"
Holmes, S. A., and W. E. Featherstone, A unified approach to the Clenshaw
summation and the recursive computation of very high degree and
order normalised associated Legendre functions, \fIJ. Geodesy\fR, 76, 279\-
299, 2002.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIplbar\fR\|(1), \fIplbar_d1\fR\|(1), \fIplmbar\fR\|(1), \fIplon\fR\|(1), \fIplon_d1\fR\|(1), \fIplmon\fR\|(1), \fIplmon_d1\fR\|(1), \fIplschmidt\fR\|(1), \fIplschmidt_d1\fR\|(1), \fIplmschmidt\fR\|(1), \fIplmschmidt_d1\fR\|(1), \fIplegendre\fR\|(1), \fIplegendre_d1\fR\|(1), \fIplegendrea\fR\|(1), \fIplegendrea_d1\fR\|(1), \fIplmindex\fR\|(1)
.PP
<http://www.ipgp.jussieu.fr/~wieczor/SHTOOLS/SHTOOLS.html>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2005\-2008 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


spherical_splines/SHTOOLS/src/PlmBar_d1.f95
subroutine PlmBar_d1(p, dp, lmax, z, csphase, cnorm)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	This function evalutates all of the normalized associated Legendre
!	functions up to degree lmax. The functions are initially scaled by 
!	10^280 sin^m in order to minimize the effects of underflow at large m 
!	near the poles (see Holmes and Featherstone 2002, J. Geodesy, 76, 279-299). 
!	On a Mac OSX system with a maximum allowable double precision value of 
!	2.225073858507203E-308 the scaled portion of the algorithm will not overflow 
!	for degrees less than or equal to 2800.
!
!	For each value of m, the rescaling factor is computed as rescalem=rescalem*sin(theta), 
!	with the intial value of rescalem being equal to 1/scalef (which is here equal 
!	to 10^280). This will gradually reduce this huge number to a tiny number, and will 
!	ultimately underflow. In order to prevent this underflow, when rescalem becomes less than
!	10^(-280), the subsequent rescaling factors of sin(theta) will be directly applied to Plm, and then this
!	number will be multipled by the old value of rescalem.
!
!	Temporary variables in saved in an allocated array. In order to explicitly deallocate this
!	memory, call this routine with a spherical harmonic degree of -1.
!
!	Calling Parameters:
!		OUT
!			p:		A vector of all associated Legendgre polynomials evaluated at 
!					z up to lmax. The lenght must by greater or equal to (lmax+1)*(lmax+2)/2.
!			dp:		A vector of all first derivatives of the normalized Legendgre polynomials evaluated at 
!					z up to lmax with dimension (lmax+1).
!		IN
!			lmax:		Maximum spherical harmonic degree to compute.
!			z:		cos(colatitude) or sin(latitude).
!		OPTIONAL (IN)
!			csphase:	1: Do not include the phase factor of (-1)^m
!					-1: Apply the phase factor of (-1)^m.
!			cnorm:		0: Use real normalization.
!					1: Use complex normalization.
!
!	Notes:
!	
!	1.	The employed normalization is the "geophysical convention." The integral of
!		(plm*cos(m theta))**2 or (plm*sin (m theta))**2 over all space is 4 pi.
!	2.	The integral of plm**2 over (-1,1) is 2 * (2 - delta(0,m)). If CNORM=1, then
!		this is equal to 2.
!	3.	The index of the array p corresponds to l*(l+1)/2 + m + 1. As such
!		the array p should be dimensions as (lmax+1)*(lmax+2)/2 in the 
!		calling routine.
!	4. 	Derivatives are calculated using the unnormalized identities
!			P'l,m = ( (l+m) Pl-1,m - l z Plm ) / (1-z**2)	(for l>m), and
!			P'll = - l z Pll / (1-z**2)	(for l=m).
!	5.	The derivative is not defined at z=+-1 for all m>0, and is therefore not
!		calculated here.
!	6.	The default is to exlude the Condon-Shortley phase of (-1)^m.
!
!
!	Dependencies:	CSPHASE_DEFAULT
!
!	Written by Mark Wieczorek September 25, 2005.
!
!	April 19, 2008: Added CNORM optional parameter compute complex normalized functions.
!
!	Copyright (c) 2008, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!	
	
	use SHTOOLS, only: CSPHASE_DEFAULT

	implicit none
	integer, intent(in) ::	lmax
	real*8, intent(out) ::	p(:), dp(:)
       	real*8, intent(in) ::	z
       	integer, intent(in), optional :: csphase, cnorm
       	real*8 ::	pm2, pm1, pmm, plm, rescalem, phase, u, scalef
      	real*8, allocatable, save ::	f1(:), f2(:), sqr(:)
      	integer ::	k, kstart, m, l, sdim, astat(3)
      	integer, save ::	lmax_old = 0

	if (lmax == -1) then
		if (allocated(sqr)) deallocate(sqr)
		if (allocated(f1)) deallocate(f1)
		if (allocated(f2)) deallocate(f2)
		lmax_old = 0
		return
	endif

      	sdim = (lmax+1)*(lmax+2)/2
      	
	if (size(p) < sdim) then 
		print*, "Error --- PlmBar_d1"
     		print*, "P must be dimensioned as (LMAX+1)*(LMAX+2)/2 where LMAX is ", lmax
     		print*, "Input array is dimensioned ", size(p)
     		stop
     	elseif (size(dp) < sdim) then 
		print*, "Error --- PlmBar_d1"
     		print*, "DP must be dimensioned as (LMAX+1)*(LMAX+2)/2 where LMAX is ", lmax
     		print*, "Input array is dimensioned ", size(dp)
     		stop
     	elseif (lmax < 0) then 
     		print*, "Error --- PlmBar_d1"
     		print*, "LMAX must be greater than or equal to 0."
     		print*, "Input value is ", lmax
     		stop
     	elseif(abs(z) > 1.0d0) then
     		print*, "Error --- PlmBar_d1"
     		print*, "ABS(Z) must be less than or equal to 1."
     		print*, "Input value is ", z
     		stop
     	elseif (abs(z) == 1.0d0) then
     		print*, "Error --- PlmBar_d1"
     		print*, "Derivative can not be calculated at Z = 1 or -1."
     		print*, "Input value is ", z
     		stop
     	endif
     	
     	
     	if (present(csphase)) then
     		if (csphase == -1) then
     			phase = -1.0d0
     		elseif (csphase == 1) then
     			phase = 1.0d0
     		else
     			print*, "PlmBar_d1 --- Error"
     			print*, "CSPHASE must be 1 (exclude) or -1 (include)."
     			stop
     		endif
     	else
     		phase = dble(CSPHASE_DEFAULT)
     	endif
     		
	scalef = 1.0d-280


	if (lmax /= lmax_old) then
		
		if (allocated(sqr)) deallocate(sqr)
		if (allocated(f1)) deallocate(f1)
		if (allocated(f2)) deallocate(f2)
		
		allocate(sqr(2*lmax+1), stat=astat(1))
		allocate(f1((lmax+1)*(lmax+2)/2), stat=astat(2))
		allocate(f2((lmax+1)*(lmax+2)/2), stat=astat(3))
		
		if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /= 0) then
			print*, "PlmBar_d1 --- Error"
			print*, "Problem allocating arrays SQR, F1 and F2", astat(1), astat(2), astat(3)
			stop
		endif

		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		!	Precompute square roots of integers that are used several times.
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
		do l=1, 2*lmax+1
			sqr(l) = sqrt(dble(l))
		enddo

		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		! 	Precompute multiplicative factors used in recursion relationships
		! 		Plmbar(l,m) = x*f1(l,m)*Plmbar(l-1,m) - Plmbar(l-2,m)*f2(l,m)
		!		k = l*(l+1)/2 + m + 1
		!	Note that prefactors are not used for the case when m=l and m=l-1,
		!	as a different recursion is used for these two values.
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
		k = 3
	
		do l=2, lmax, 1
			k = k + 1
			f1(k) = sqr(2*l-1) * sqr(2*l+1) / dble(l)
			f2(k) = dble(l-1) * sqr(2*l+1) / sqr(2*l-3) / dble(l)
			do m=1, l-2
				k = k+1
				f1(k) = sqr(2*l+1) * sqr(2*l-1) / sqr(l+m) / sqr(l-m)
                		f2(k) = sqr(2*l+1) * sqr(l-m-1) * sqr(l+m-1) &
                  			 / sqr(2*l-3) / sqr(l+m) / sqr(l-m) 
			enddo
			k = k + 2
		enddo
		
		lmax_old = lmax
		
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!	
	!	Calculate P(l,0). These are not scaled.
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

	u = sqrt((1.0d0-z)*(1.0d0+z)) ! sin(theta)

      	pm2  = 1.0d0	
      	p(1) = 1.0d0
      	dp(1) = 0.0d0
      	
      	if (lmax == 0) return
      	
      	pm1  = sqr(3)*z	
      	p(2) = pm1
      	dp(2) = sqr(3)
      		
	k = 2

      	do l = 2, lmax, 1
         	k = k+l
         	plm = f1(k)*z*pm1-f2(k)*pm2
         	p(k) = plm
         	dp(k) = dble(l) * ( sqr(2*l+1) / sqr(2*l-1)  * &
      			pm1 - z * plm ) / u**2
         	pm2  = pm1
         	pm1  = plm
      	enddo

	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate P(m,m), P(m+1,m), and P(l,m)
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!	

	if (present(cnorm)) then
		if (cnorm == 1) then
			pmm = scalef
		else
			pmm = sqr(2)*scalef
		endif
	else
      		pmm = sqr(2)*scalef
      	endif
	
      	rescalem = 1.0d0/scalef
      	kstart = 1

      	do m = 1, lmax - 1, 1
      		
      		rescalem = rescalem*u

		! Calculate P(m,m)
        	kstart = kstart+m+1
         
         	pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
        	p(kstart) = pmm*rescalem
        	dp(kstart) = -dble(m) * z * p(kstart) / u**2
        	pm2 = pmm

		! Calculate P(m+1,m)
		k = kstart+m+1
	   	pm1 = z * sqr(2*m+3) * pmm
	    	p(k) = pm1*rescalem
	    	dp(k) =  ( sqr(2*m+3) * p(k-m-1) - z * dble(m+1) * p(k)) / u**2

		! Calculate P(l,m)
               	do l = m+2, lmax, 1
               		k = k+l
                  	plm  = z*f1(k)*pm1-f2(k)*pm2
                  	p(k) = plm*rescalem
                  	dp(k) = ( sqr(2*l+1) * sqr(l-m) * sqr(l+m) / sqr(2*l-1) * &
      				p(k-l) - z * dble(l) * p(k)) / u**2
                  	pm2  = pm1
                  	pm1  = plm
               	enddo
              
      	enddo
      	
      	! Calculate P(lmax,lmax)
      
	rescalem = rescalem*u	
   	
        kstart = kstart+m+1
        pmm = phase * pmm * sqr(2*lmax+1) / sqr(2*lmax)
        p(kstart) = pmm*rescalem
        dp(kstart) = -dble(lmax) * z * p(kstart) / u**2
      		
end subroutine PlmBar_d1

